### DISCUSSION TITLE
Cohort data for different age bands

### DISCUSSION THREAD

**Discussion by YatishRam** [opensafely/documentation#133] [Q&A] [answered]
We report on vaccination under different age group. Something like 

Age Group | Total Patient | Patients vaccinated
-- | -- | --
0 - under 16 | 100 | 50
16 - under 30 | 200 | 100
30 - under   40 | 250 | 75

I have created a study definition with population satisfying a certain age group. 
```
    population = patients.satisfying("age < 16 ",                                     
       age=patients.age_as_of(index_date)),
```
Is it possible to have a single study definition with population targeting different age groups or do I need to create multiple study definitions and extract cohort data separately?

**Comment by wjchulme** [comment, marked as answer]
The typical way to analyse different age groups (or any other category) seperately would be to define an age group variable (either within the study definition, or after you have imported the extracted data for analysis) and then stratify or group your analysis on that variable. 

To create an age group variable within the study definition you can use the [`patients.categorised_as`](https://docs.opensafely.org/en/latest/study-def-variables/#cohortextractor.patients.categorised_as) function as follows (assuming `age` is defined elsewhere in the study definition):

```py
age_group = patients.categorised_as(
        {
            "0": "DEFAULT",
            "16-59": """ age >= 16 AND age < 60""",
            "60-69": """ age >= 60 AND age < 70""",
            "70-79": """ age >= 70 AND age < 80""",
            "80+": """ age >=  80 AND age < 120""", 
        },
        return_expectations={
            "rate": "universal",
            "category": {
                "ratios": {
                    "16-59": 0.5,
                    "60-69": 0.125,
                    "70-79": 0.25,
                    "80+": 0.125,
                }
            },
        },
    ),
```


For the specific usecase you describe, you might be interested in using [`measures`](https://docs.opensafely.org/en/latest/measures/). You could take advantage of the `group_by` functionality to group your denominator (number of patients) and your numerator (number of patients vaccinated) by age group. You can do this for either for a single time period or over multiple time periods.

**Reply by sebbacon** [reply]
@YatishRam please mark this answer as "accepted" if it answers your question. Thanks!

**Reply by YatishRam** [reply]
> The typical way to analyse different age groups (or any other category) seperately would be to define an age group variable (either within the study definition, or after you have imported the extracted data for analysis) and then stratify or group your analysis on that variable.
> 
> To create an age group variable within the study definition you can use the [`patients.categorised_as`](https://docs.opensafely.org/en/latest/study-def-variables/#cohortextractor.patients.categorised_as) function as follows (assuming `age` is defined elsewhere in the study definition):
> 
> ```python
> age_group = patients.categorised_as(
>         {
>             "0": "DEFAULT",
>             "16-59": """ age >= 16 AND age < 60""",
>             "60-69": """ age >= 60 AND age < 70""",
>             "70-79": """ age >= 70 AND age < 80""",
>             "80+": """ age >=  80 AND age < 120""", 
>         },
>         return_expectations={
>             "rate": "universal",
>             "category": {
>                 "ratios": {
>                     "16-59": 0.5,
>                     "60-69": 0.125,
>                     "70-79": 0.25,
>                     "80+": 0.125,
>                 }
>             },
>         },
>     ),
> ```
> 
> For the specific usecase you describe, you might be interested in using [`measures`](https://docs.opensafely.org/en/latest/measures/). You could take advantage of the `group_by` functionality to group your denominator (number of patients) and your numerator (number of patients vaccinated) by age group. You can do this for either for a single time period or over multiple time periods.

Measures do the job. But it generates one csv per measure, that's problem because we have over hundred measures.
Is there a way to stitch it all together into one csv? Basically have all the measures in one file.

**Reply by wjchulme** [reply]
To clarify, it generates a set of CSVs _for each measure and each time period_ and also combines these CSVs into a single CSV _for each measure_. Perhaps this latter file is what you need?

However, if you're talking about combining CSVs across different measures then this is complicated by the fact that the column names don't neessarily agree between measures. For example if you have two measures, proportion vaccinated by age group and proportion dying by age group, then this will result in two (combined) CSV files, with column names `[age_group, vaccinated, population, value, date]` for the first measure, and `[age_group, died, population, value, date]` for the second measure. In practice the first three column names could all be different. 

To stack multiple measures together into one file, you'd need to standardise the column names, add an extra row that identifies the measure, and then bind them together. This could be automated in a small python or R script.

**Reply by YatishRam** [reply]
I got it confused because all my measures had the same denominator and the grouping. So it seemed odd to get one file per measure.
But as you said all measures can essentially be different column and grouping.
